sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/exemplar/AbstractFixedSizeExemplarReservoir.java

```diff
@@ -0,0 +1,152 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.sdk.metrics.exemplar;
+
+import io.opentelemetry.api.common.AttributeKey;
+import io.opentelemetry.api.common.Attributes;
+import io.opentelemetry.api.common.AttributesBuilder;
+import io.opentelemetry.api.trace.Span;
+import io.opentelemetry.context.Context;
+import io.opentelemetry.sdk.common.Clock;
+import io.opentelemetry.sdk.metrics.data.DoubleExemplar;
+import io.opentelemetry.sdk.metrics.data.Exemplar;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Set;
+import javax.annotation.Nullable;
+
+/**
+ * Base implementation for fixed-size reservoir sampling of Exemplars.
+ *
+ * <p>Additionally this implementation ONLY exports double valued exemplars.
+ */
+abstract class AbstractFixedSizeExemplarReservoir implements ExemplarReservoir {
+  private final Clock clock;
+  private final ReservoirCell[] storage;
+
+  /**
+   * Instantiates an exemplar reservoir of fixed size.
+   *
+   * @param clock The clock to use when annotating measurements with time.
+   * @param size The number of exemplars to preserve.
+   */
+  AbstractFixedSizeExemplarReservoir(Clock clock, int size) {
+    this.clock = clock;
+    this.storage = new ReservoirCell[size];
+    for (int i = 0; i < size; ++i) {
+      this.storage[i] = new ReservoirCell();
+    }
+  }
+
+  protected final int maxSize() {
+    return storage.length;
+  }
+  /**
+   * Determines the sample bucket for a given measurement.
+   *
+   * @return The bucket to sample into or -1 for no sampling.
+   */
+  protected abstract int bucketFor(double value, Attributes attributes, Context context);
+  /** Callback to reset any local state after a {@link #collectAndReset} call. */
+  protected void reset() {}
+
+  @Override
+  public final void offerMeasurement(long value, Attributes attributes, Context context) {
+    offerMeasurement((double) value, attributes, context);
+  }
+
+  @Override
+  public final void offerMeasurement(double value, Attributes attributes, Context context) {
+    int bucket = bucketFor(value, attributes, context);
+    if (bucket != -1) {
+      this.storage[bucket].offerMeasurement(value, attributes, context);
+    }
+  }
+
+  @Override
+  public final List<Exemplar> collectAndReset(Attributes pointAttributes) {
+    // Note: we are collecting exemplars from buckets piecemeal, but we
```

Comment:
This probably should be on a javadoc, though I guess not here since it's a package private class

Reply:
It's a bit of a concurrency/implementation detail.   Nominally subclasses shouldn't care/notice.  If it shows up, it's a bit of an (acceptable) bug.  I think fixing this to be perfect would impact performance way too much to be worth it.  It's possible we can fix this with a timestamp solution too.  The odds this happens is so esoterically low, I don't think it's worth loosing performance BUT would be good if someone else agrees with that assessment of risk vs. reward.